parlai/tasks/coco_caption/agents.py

```diff
@@ -0,0 +1,301 @@
+
+        if shared and 'ques' in shared:
+            # another instance was set up already, just reference its data
+            if 'annotation' in shared:
+                self.annotation = shared['annotation']
+            self.image_loader = shared['image_loader']
+        else:
+            # need to set up data from scratch
+            test_info_path, annotation_path, self.image_path = _path(opt, version)
+            self._setup_data(test_info_path, annotation_path)
+            self.image_loader = ImageLoader(opt)
+
+        self.reset()
+
+    def reset(self):
+        super().reset()  # call parent reset so other fields can be set up
+        self.example = None  # set up caching fields
+        self.next_example()  # call this once to get the cache moving
+
+    def num_examples(self):
+        # We only have annotations for the train and val sets, so for the test
+        # set we need to determine how many images we have.
+        if not self.datatype.startswith('test'):
+            return len(self.annotation['annotations'])
+        else:
+            return len(self.test_info['images'])
+
+    def num_episodes(self):
+        return self.num_examples()
+
+    def submit_load_request(self, image_id):
+        img_path = self.image_path + '%012d.jpg' % (image_id)
+        self.data_loader.request_load(self.receive_data, self.image_loader.load, (img_path,))
+
+    def get(self, episode_idx, entry_idx=0):
+        action = {
+            'text': self.ques,
+            'episode_done': True
+        }
+
+        if not self.datatype.startswith('test'):
+            # test set annotations are not available for this dataset
+            anno = self.annotation['annotations'][episode_idx]
+            action['labels'] = [anno['caption']]
+            action['image_id'] = anno['image_id']
+        else:
+            action['image_id'] = self.test_info['images'][episode_idx]['id']
+
+        return action
+
+    def next_example(self):
+        """Returns the next example from this dataset after starting to queue
+        up the next example.
+        """
+        ready = None
+        # pull up the currently queued example
+        if self.example is not None:
+            if self.image_mode != 'none':
+                # move the image we loaded in the background into the example
+                image = self.data_queue.get()
+                self.example['image'] = image
+            ready = (self.example, self.epochDone)
+        # get the next base example: super().next_example() calls self.get()
+        self.example, self.epochDone = super().next_example()
+        if self.image_mode != 'none' and 'image_id' in self.example:
+            # load the next image in the background
+            image_id = self.example['image_id']
+            self.submit_load_request(image_id)
+        # return the previously cached example
+        return ready
+
+    def share(self):
+        shared = super().share()
+        shared['ques'] = QUESTION
```

Comment:
We don't need to share `QUESTION` - ``share`` is mostly for larger data items for which multiple copies don't make sense (e.g. the annotations)